### ISSUE TITLE
npm says lerna-created node_modules/.bin/file.js symlinks in package directories 'isn't under npm's control' and crashes/non-zero exits

### ISSUE THREAD

**Issue by wbern** [lerna/lerna#1812]
I realize this may be related to npm, but I want to flag this as an issue with lerna still. I very much appreciate the symlinking behavior of the bin files and the hoist mechanism, so I want to make sure to show this issue, hoping for it to be fixed.

## Expected Behavior

I'd like to be able to run `npm install` prior to all `lerna bootstrap --hoist` commands. This is to get lerna installed as a local dependency as well as to bootstrap lerna in one command, aka `npm install`.

## Current Behavior

Because I have some `"bin"` property defined in one of my packages, I get the following error on doing `npm install`.

```
npm ERR! path /home/user/Repos/frontend-project/node_modules/package-vue-chartjs/node_modules/.bin/get-require-path
npm ERR! code EEXIST
npm ERR! Refusing to delete /home/user/Repos/frontend-project/node_modules/package-vue-chartjs/node_modules/.bin/get-require-path: containing path /home/user/Repos/frontend-project/node_modules/package-vue-chartjs/node_modules/app-manager-shared isn't under npm's control
npm ERR! File exists: /home/user/Repos/frontend-project/node_modules/package-vue-chartjs/node_modules/.bin/get-require-path
npm ERR! Move it away, and try again.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/user/.npm/_logs/2018-12-05T07_46_50_546Z-debug.log
[08:46:50] 'micro-app-generator-shared:default' errored after 21 s Command `npm install --save-dev micro-apps/components/my-package && npm install` exited with code 1
/home/user/Repos/frontend-project/node_modules/orchestrator/index.js:153
			throw err;
			^

Error: Command `npm install --save-dev package` exited with code 1
    at ChildProcess.handleSubShellExit (/home/user/Repos/frontend-project/node_modules/gulp-run/command.js:166:13)
    at Object.onceWrapper (events.js:317:30)
    at emitTwo (events.js:126:13)
    at ChildProcess.emit (events.js:214:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! MyBusiness@1.0.0 create-micro-app: `cross-env generation_path='./micro-apps/' slush micro-app-generator-shared`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the MyBusiness@1.0.0 create-micro-app script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/user/.npm/_logs/2018-12-05T07_46_50_600Z-debug.log
```

## Possible Solution

Workaround is to run this in the `preinstall` hook of package.json, but unfortunately that hook won't run when you attempt to install a package, and it's also not very pretty.

```
"preinstall": "find ./micro-apps -name .bin -type d -exec rm -rf {} +",
```

## Steps to Reproduce (for bugs)

1. Add a package A
2. Add a a "bin" property to that package.json file of that package A
3. Add a new package B,  specify dependency to package A
4. Do `npm install`, `lerna bootstrap --hoist`, then `npm install` again.

<details><summary>lerna.json</summary><p>

```json
{
    "version": "independent",
    "packages": [
      "micro-apps/*/*",
      "shared-apps/*"
    ],
    "command": {
      "publish": {
        "ignoreChanges": [
          "ignored-file",
          "*.md"
        ],
        "independent": true
      },
      "bootstrap": {
        "npmClientArgs": ["--no-package-lock"] 
      }
    },
    "useWorkspaces": false
  }
```
</p></details>

<details><summary>lerna-debug.log</summary><p>

It doesn't populate with this error.
</p></details>

## Context

I want to have `bin` executable scripts that I can share among my packages to be executed inside `package.json`. Since require() can traverse upwards to look in the root project node_modules and `package.json` cannot do this, this is my best bet to be able to execute things.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.5.1 |
| `npm --version`  | 5.6.0 |
| `node --version` | 8.11.3 |

| OS | Version |
| --- | --- |
| Ubuntu | 18.04.1 |

**Comment by wbern** [comment]
Are you saying I should do my workaround in that script instead?

**Comment by evocateur** [comment]
No, a lifecycle to "hook" npm install with bootstrap. Or run `npx lerna link convert` once and stop bootstrapping altogether.

**Comment by wbern** [comment]
While this definitely solved my problem of how to hook together `npm install` and lerna's bootstrapping mechanism, (I was using postinstall before which was.. not as good), the original issue in this subject is still hard to avoid sometimes. It can also happen if a user decides to do `npm install` inside a micro-app directory. When going back to the root project dir and doing `npm install`, they get complaints from npm by the bin directory having files it doesn't "own".

My work-around is fine for now though. Closing

**Comment by wbern** [comment]
@evocateur I'm running into random situations where the "prepare" script gets called after lerna bootstrap --hoist.. It doesn't always happen but as I move packages around it now seems to be happening all the time.

Do you have any reference to your source code when "prepare" is being called? I don't want install loops.

**Comment by evocateur** [comment]
https://github.com/lerna/lerna/commit/ea9dbbee12f9df86789f8739efc49fc612d9378d should have fixed the recursive `lerna bootstrap` from the `prepare` lifecycle (as well as `prepublish`, `preinstall`, `install`, and `postinstall` in the root). It was released in v3.9.1, but you might as well use v3.10.1 (latest).

All of those lifecycles are run during any leaf package `npm install` (if anything is not hoisted), as well as the root. They are also run in topological order [here](https://github.com/lerna/lerna/blob/51fe2c4ae4b76c1fadf31797bee178c20c2ec3ce/commands/bootstrap/index.js#L152-L187). (I don't like that they're being run twice, but I can't figure out a better way to ensure they are always run in a consistent fashion)

**Comment by wbern** [comment]
@evocateur thanks!
